src/messenger.rs

```diff
@@ -495,6 +504,107 @@ fn match_versions(range_a: ApiVersionRange, range_b: ApiVersionRange) -> Option<
     }
 }
 
+/// Helper that ensures that a request is removed when a request is cancelled before it was actually sent out.
+struct CleanupRequestStateOnCancel {
+    state: Arc<Mutex<MessengerState>>,
+    correlation_id: i32,
+    message_sent: bool,
+}
+
+impl CleanupRequestStateOnCancel {
+    /// Create new helper.
+    ///
+    /// You must call [`message_sent`](Self::message_sent) when the request was sent.
+    fn new(state: Arc<Mutex<MessengerState>>, correlation_id: i32) -> Self {
+        Self {
+            state,
+            correlation_id,
+            message_sent: false,
+        }
+    }
+
+    /// Request was sent. Do NOT clean the state any longer.
+    fn message_sent(mut self) {
+        self.message_sent = true;
+    }
+}
+
+impl Drop for CleanupRequestStateOnCancel {
+    fn drop(&mut self) {
+        if !self.message_sent {
+            if let MessengerState::RequestMap(map) = self.state.lock().deref_mut() {
+                map.remove(&self.correlation_id);
+            }
+        }
+    }
+}
+
+pin_project! {
+    struct CancellationSafeFuture<F>
+    where
+        // https://github.com/taiki-e/pin-project-lite/issues/2#issuecomment-745190950
+        F: Future,
+        F: Send,
+        F: 'static,
+    {
+        // Mark if the inner future finished. If not, we must spawn a helper task on drop.
+        done: bool,
+
+        // Inner future.
+        //
+        // Wrapped in an `Option` so we can extract it during drop. Inside that option however we also need a pinned
+        // box because once this wrapper is polled, it will be pinned in memory -- even during drop. Now the inner
+        // future does not necessarily implement `Unpin`, so we need a heap allocation to pin it in memory even when we
+        // move it out of this option.
+        #[pin]
```

Comment:
As BoxFuture is Unpin, I don't think you need to use pin_project at all

Reply:
done